go/types/subtype.go

```diff
@@ -105,73 +120,108 @@ func isSubtype(requiredType, concreteType *Type, allowExtraStructFields bool, pa
+				// Concrete contains extra field
+				hasExtra = true
+				j++
 			}
-
-			j++
 		}
 
 		for i < requiredDesc.Len() {
 			// Fields in required not in concrete
 			if !requiredDesc.fields[i].Optional {
-				return false
+				hasExtra = true
+				return false, hasExtra
 			}
 			i++
 		}
 
-		if j < concreteDesc.Len() && !allowExtraStructFields {
-			return false
-		}
-
-		return true
-
+		hasExtra = hasExtra || j < concreteDesc.Len()
+		return true, hasExtra
 	}
 
 	panic("unreachable")
 }
 
 // compoundSubtype is called when comparing the element types of two compound types. This is the only case
 // where a concrete type may have be a union type.
-func compoundSubtype(requiredType, concreteType *Type, allowExtraStructFields bool, parentStructTypes []*Type) bool {
+func compoundSubtype(requiredType, concreteType *Type, hasExtra bool, parentStructTypes []*Type) (bool, bool) {
 	// If the concrete type is a union then all the types in the union must be subtypes of the required typ. This also means that a compound type with an empty union is going to be a subtype of all compounds, List<> is a subtype of List<T> for all T.
 	if concreteType.TargetKind() == UnionKind {
 		for _, ct := range concreteType.Desc.(CompoundDesc).ElemTypes {
-			if !isSubtype(requiredType, ct, allowExtraStructFields, parentStructTypes) {
-				return false
+			isSub, hasExtra1 := isSubtype(requiredType, ct, hasExtra, parentStructTypes)
+			if !isSub {
+				return false, hasExtra1
 			}
 		}
-		return true
+		return true, hasExtra
 	}
-	return isSubtype(requiredType, concreteType, allowExtraStructFields, parentStructTypes)
+	return isSubtype(requiredType, concreteType, hasExtra, parentStructTypes)
 }
 
-// IsValueSubtypeOf returns whether a value is a subtype of a type.
-func IsValueSubtypeOf(v Value, t *Type) bool {
+// IsValueSubtypeOf returns whether a value is a subtype of a type and whether
+// it has extra fields. See comment above on function: isSubtypeTopLevel().
+// IsValueSubtypeOf returns two values:
+//   isSub - which indicates whether concreteType is a subtype of requiredType.
+//   hasExtra - which indicates whether concreteType has additional fields.
+//     This field has no meaning if IsSub is false.
+//
+// For example, given the following data:
+//   type1 := struct S {               v := Struct S1 {
+//       a Number | string                 a: "hello"
+//       b ?int                            b: 2
+//   }                                 }
+// IsValueSubtypeOf(v, type1) would return isSub == true, and hasExtra == false
+//
+// And given these types:
+//   type2 := struct S {               v := Struct S1 {
+//       a Number | string                 a: "hello"
+//       b ?int                            b: 2
+//   }                                     c: "hello again"
+//                                     }
+// IsValueSubtypeOf(v, type1) would return isSub == true, and hasExtra == true
+func IsValueSubtypeOf(v Value, t *Type) (bool, bool) {
```

Comment:
Don't return two booleans here. Just add another function that does that. It is very strange to ever want to care about this strange hasExtra thing. It is not part of subtype it is just a strange thing that most people should never have to run into.